### PR TITLE
[dtensor] refactor einop_rule and mm ops to handle reshard uniformly

### PR DESCRIPTION
**Pull request by wanchaol**
This PR refactors einop_rule and mm ops (namely addmm/baddbmm) to handle reshard suggestions uniformly. Changes:
1. merge the partial_linearity logic with needs_reshard logic, so that partial linearity is just once case that needs reshard in einop rule
2. refactor mm ops to handle reshard suggestions uniformly
3. add backward tests for bmm and baddbmm op
4. enable triggering partial all_reduce when linearity=False

### CODE REVIEW

File: test/spmd/tensor/test_matrix_ops.py [hunk 1]
```diff
@@ -132,19 +132,26 @@ def test_t(self):
 
     # baddbmm introduces nan occasionally on CPU: https://github.com/pytorch/pytorch/issues/80588
     @with_comms
-    @skip_if_lt_x_gpu(TEST_GPU_NUM)
-    def test_sharded_baddbmm(self):
+    # @skip_if_lt_x_gpu(TEST_GPU_NUM)
```
Comment (aazzolini) [inline test/spmd/tensor/test_matrix_ops.py:135]: Do we want to un-comment this before landing?

File: test/spmd/tensor/test_matrix_ops.py [hunk 2]
```diff
@@ -155,15 +162,25 @@ def test_placement_comb(
...
                 DTensor,
                 torch.baddbmm(
                     tensor_dt, batch_1_dt, batch_2_dt, beta=beta, alpha=alpha
                 ),
             ).redistribute(device_mesh, [Replicate()])
-            assert not torch.isnan(local_result).any()
-            assert not torch.isnan(new_dt.to_local()).any()
-            self.assertEqual(new_dt.to_local(), local_result)
+            dist_local_res = dist_res.to_local()
+            # assert not torch.isnan(local_result).any()
```
Comment (aazzolini) [inline test/spmd/tensor/test_matrix_ops.py:172]: REmove?

File: spmd/tensor/ops/common_rules.py [hunk 3]
```diff
@@ -170,8 +132,19 @@ def merge_sharding(dim: str, a: int, b: int) -> int:
                 )
                 assert dim_to_size[dim] == input_spec.shape[idx]
 
+    if pending_sums_counter and not linearity:
+        raise RuntimeError("Cannot do generic op on a tensor with partial sums")
```
Comment (aazzolini) [inline spmd/tensor/ops/common_rules.py:136]: Can we handle this instead of raising?

**Review by aazzolini — APPROVED** [review]

File: spmd/tensor/ops/common_rules.py (hunk 3 above)
Comment (wanchaol) [reply spmd/tensor/ops/common_rules.py:136]: good point, let me add this directly

**Pull request merged**